### DISCUSSION TITLE
Support for Login V2 base theme

### DISCUSSION THREAD

**Discussion by topdev-spetermann** [keycloakify/keycloakify#729] [General]
Hello everyone, hello @garronej,

currently in generated theme.properties file for login theme there is generated the line
`parent=keycloak`
at the beginning, which is not overridable i think.
This means, that the old login theme from parent keycloak is always used as the base theme with it's css files and the very old Patternfly version.
With Keycloak 26 the "Login V2" theme was introduced in the theme keycloak.v2. The old theme keycloak is deprecated now and will be removed in the future. 

Is there already a plan to support this new login theme as parent theme in keycloakify? And is it already possible to define a timeline for this?

**Comment by garronej** [comment]
Hello @topdev-spetermann,

We'll eventually update, but not until Keycloak v2 is stable and officially becomes the base theme.

> `parent=keycloak`

This can be ignored—it could be `parent=base` and everything would work the same. The Keycloakify components are a complete React port of the FTL UI. So, it's not just a matter of updating this; many other things need to be updated, and it might introduce breaking changes.

While we want to stay reasonably up to date with the default theme, our top priority is to be feature complete (ensuring everything that works in the default theme works in the base Keycloakify UI), stable (avoiding any issues that might break people's themes when they update to a newer Keycloakify version), and backward compatible (ensuring Keycloakify themes work on every version of Keycloak).

Having the latest PatternFly version and other cosmetic updates is a low priority compared to those three aspects, as most of our user base chooses to disable PatternFly.  

If creating Keycloak theme is an integral part of your buisness and you would like to see this effort prioritize consider contributing though GitHub Sponsors.  

Best,

**Reply by topdev-spetermann** [reply]
Hello @garronej,

for my understanding it is stable: The [release notes of Keyclok 26](https://www.keycloak.org/docs/latest/release_notes/index.html#new-default-login-theme) already wrote:

> New default login theme
> 
> There is now a new version (v2) of the keycloak login theme, which provides an improved look and feel, including support for switching automatically to a dark theme based on user preferences.
> 
> The previous version (v1) is now deprecated, and will be removed in a future release.
> 
> For all new realms, keycloak.v2 will be the default login theme. Also, any existing realm that never explicitly set a login theme will be switched to keycloak.v2.

So i'm not sure what's your definition of stable.

And i'm not sure, but i'm think there would be a major difference, when parent would be base: Patternfly 3 is part of the theme.properties in theme "keycloak" and not in base. So if i'm right and there is no default theme property for including Patternfly 3, it would not be included in our theme out of the box.

It would be enough for us to have a option to override this parent on build, so we could change to the keycloak.v2 theme as parent on ourselve. Currently it seems, that the parent is hardcoded to "keycloak" and if i try to add it as extra theme property it renders the parent twice, but only the first one is evaluated.

**Reply by garronej** [reply]
> So i'm not sure what's your definition of stable.  

Ok, my bad, it's stable and we can put migrating to it in the roadmap.  

> And i'm not sure, but i'm think there would be a major difference  

Trust me, there will be no diffrence. Go ahead an try putting "base" instead of "keycloak" you'll see no diffrence.  
All Keycloakify theme are self contained and don't inherit anything from what's in Keycloak.  
We only set `parent=keycloak` because some poweruser deletes some keycloakify pages so that they can fallback to the default page.  

You can play along with https://docs.keycloakify.dev/in-depth-configuration/configuration-options/postbuild  
Or use `npx keycloakify start-keycloak` and manually edit the files in `dist_keycloak/` if you want to experiment.

**Reply by topdev-spetermann** [reply]
Thanks for the reference to the options. I missed that. We will have a look for this.

**Reply by garronej** [reply]
You're welcome but don't waste your time though. It's not possible to upgrade to v2 just on your end.

**Reply by topdev-spetermann** [reply]
I know, but if we have enough time for this, we could participate in migration with PR's from a forked repo ;). But i have to clarify this with our customer.

**Reply by mabartos** [reply]
Yes, the Login v2 theme is the default one used in Keycloak now. 

JFYI - for the upcoming **Keycloak 26.2** there will be some changes made to these login screens to make it more user-friendly. (See https://github.com/keycloak/keycloak/issues/38117)

**Keycloakify** is also a good tool to verify the layout or CSS styles are not broken for these screens. So, I'd be very interested in having the Login v2 enabled by default for Keycloakify to even confirm all screens are OK (to fulfill this issue: https://github.com/keycloak/keycloak/issues/38332).

FYI - if you find some issues with specific Login v2 screens, create an issue in the Keycloak repository which will be then attached to this issue: https://github.com/keycloak/keycloak/issues/38575.

**Reply by garronej** [reply]
Hello @mabartos,

I’d like to move forward with updating to Keycloak Login v2. This is something that needs to be done, but there are a few important considerations:

- Updating without breaking existing themes and ensuring a clear migration path requires significant effort.
- Converting FreeMarker templates into React/Angular/Svelte components is a meticulous and time-consuming task. It's being done on weekends by myself and a few community members.

Given Keycloakify’s wide adoption, could we open a discussion about Red Hat either:
- Assisting with the FreeMarker-to-React migration, or
- Providing some financial support?

To be clear, this isn’t a way to make support conditional on contribution. I'll keep maintaining Keycloakify for the forseable future but our primary goal is to maintain stability and compatibility. Keeping up with the latest Keycloak developments comes only after that.

Thanks in advance!

**Reply by garronej** [reply]
Hello @mabartos,

Thanks again for reaching out last week, I really appreciated the context on Login v2 and the heads-up on upcoming Keycloak changes.

I realize my previous message may have shifted the tone a bit too abruptly. That wasn't my intention, I was juggling a lot, and I didn't express enough how valuable your input was. It's always great to see Keycloakify being used directly by the team and to get actionable feedback like this.

I'll be working on Login v2 support ASAP and will follow up if any compatibility issues arise.

Thanks again for the engagement!

**Reply by mabartos** [reply]
@garronej Sorry for the late reply, I kinda lost it in my inbox. 

 I understand that the work on this popular project is time-consuming, and your main focus is to comply with your requirements/needs, and the latest versions development has a lower priority. My previous message was not about pushing you to start working on it, but rather like: "it'd be nice to have it as Keycloak community would appreciate it as well :)" 

> Given Keycloakify’s wide adoption, could we open a discussion about Red Hat

I understand what you mean. However, Keycloak is part of [CNCF](https://www.cncf.io/projects/keycloak/) organization(foundation) and is only sponsored by Red Hat in the sense that the majority of the development team is from that company.

I cannot say anything official on behalf of Red Hat or what the process is for supporting such projects as Keycloakify - simply I don't even know what needs to be done. IMHO, if anybody from the Keycloak team is willing to help with the migration, I think it's welcome, but should not be required on the company level, but rather as a community contribution. Even based on the recently announced transition of Red Hat middleware to IBM, the process of supporting such projects might be simpler and more easily achievable (maybe?). So, I'd suggest wait, and then we can look for some options :) Does it work for you? 

Thanks for the work on this!

**Reply by kathari00** [reply]
> **Keycloakify** is also a good tool to verify the layout or CSS styles are not broken for these screens. So, I'd be very interested in having the Login v2 enabled by default for Keycloakify to even confirm all screens are OK (to fulfill this issue: [keycloak/keycloak#38332](https://github.com/keycloak/keycloak/issues/38332)).

Hello @mabartos. I think by using [keycloak theme editor](https://keycloak-theme-editor.org) you would be able to reproduce layout issues as well as the pages are generated by compiling Freemarker with test data.

You could potentially test new development (if you update to Patternfly v6 eg) by using the CLI commands.

**Comment by garronej** [comment]
Just an update to say that @kathari00 and I have started working on it.

**Reply by mars3142** [reply]
Any progress on it? I need v2 base, because I can't use passkey anymore with the keycloakify theme.

**Reply by kathari00** [reply]
Hello, even though V2 is not published yet, there was a minor update to support passkeys in keycloakify V1. One way or the other you can always extend keycloakify to your liking.

**Reply by garronej** [reply]
@mars3142 While login v2 is still in progress, the React version of the current template has already been updated to support passkeys.

See:
- https://github.com/keycloakify/keycloakify/issues/990
- https://github.com/keycloakify/keycloakify/issues/918

If you're still running into issues with passkeys on the current template, please open a new issue with reproduction steps (Keycloak version, Keycloakify version, and any relevant logs/screenshots).  

At the moment, I don't think there is any feature missing from the current template, at least in React. It's mostly that the default style is lacking behind.  
But we have an alternative base look build on with Shadecn/Tailwind: https://docs.keycloakify.dev/starter-themes/shadcn-ui-tailwind

**Reply by mars3142** [reply]
@garronej You wrote "in React", but I use Svelte. That would explain, why I can't see it.

I use (as quick fix) the keycloak theme editor, which helps for the MVP of my project.

**Reply by garronej** [reply]
@luca-peruzzo,  
If you have the bandwidth to port this update to the Svelte version of Login v1, that would be super helpful.  
That said, I know you’ve already put a lot of work into v2, and I still haven’t had the chance to review it because I’m not finished with the React implementation of v2 yet.
